lib/experimental/fonts-library/class-wp-font-family.php

```diff
@@ -0,0 +1,421 @@
+	 */
+	private function download_asset( $src, $filename ) {
+		$file_path = path_join( WP_FONTS_DIR, $filename );
+
+		// Checks if the file to be downloaded has a font mime type.
+		if ( ! WP_Font_Family_Utils::has_font_mime_type( $file_path ) ) {
+			return false;
+		}
+
+		// Include file with download_url() if function doesn't exist.
+		if ( ! function_exists( 'download_url' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/file.php';
+		}
+
+		// Downloads the font asset or returns false.
+		$temp_file = download_url( $src );
+		if ( is_wp_error( $temp_file ) ) {
+			return false;
+		}
+
+		// Moves the file to the fonts directory or return false.
+		$renamed_file = rename( $temp_file, $file_path );
+		// Cleans the temp file.
+		@unlink( $temp_file );
+
+		if ( ! $renamed_file ) {
+			return false;
+		}
+
+		// Returns the relative path to the downloaded font asset to be used as font face src.
+		return WP_Fonts_Library::get_relative_fonts_path() . $filename;
+	}
+
+	/**
+	 * Move an uploaded font face asset from temp folder to the wp fonts directory.
+	 *
+	 * This is used when uploading local fonts.
+	 *
+	 * @param array $font_face Font face to download.
+	 * @param array $file Uploaded file.
+	 * @return array|bool New font face with all assets downloaded and referenced in the font face definition or false is something failed
+	 */
+	private function move_font_face_asset( $font_face, $file ) {
+		$new_font_face = $font_face;
+		$filename      = WP_Font_Family_Utils::get_filename_from_font_face( $font_face, $file['name'] );
+		$filepath      = path_join( WP_FONTS_DIR, $filename );
+
+		// Remove the uploaded font asset reference from the font face definition because it is no longer needed.
+		unset( $new_font_face['file'] );
+
+		// If the filepath has not a font mime type, we don't move the file and return the font face definition without src to be ignored later.
+		if ( ! WP_Font_Family_Utils::has_font_mime_type( $filepath ) ) {
+			return $new_font_face;
+		}
+
+		// Move the uploaded font asset from the temp folder to the wp fonts directory.
+		$file_was_moved = move_uploaded_file( $file['tmp_name'], $filepath );
+
+		if ( $file_was_moved ) {
+			// If the file was successfully moved, we update the font face definition to reference the new file location.
+			$new_font_face['src'] = WP_Fonts_Library::get_relative_fonts_path() . $filename;
+		}
+
+		return $new_font_face;
+	}
+
+	/**
+	 * Sanitizes the font family data using WP_Theme_JSON.
+	 *
+	 * @return array A sanitized font family defintion.
+	 */
+	private function sanitize() {
+		// Creates the structure of theme.json array with the new fonts.
+		$fonts_json = array(
+			'version'  => '2',
+			'settings' => array(
+				'typography' => array(
+					'fontFamilies' => array( $this->data ),
+				),
+			),
+		);
+		// Creates a new WP_Theme_JSON object with the new fonts to leverage sanitization and validation.
+		$theme_json     = new WP_Theme_JSON( $fonts_json );
```

Comment:
I think we might want to use WP_Theme_JSON_Gutenberg while this is still experimental and not in Core. Documentantion on this is sparse, but here's the PR introducing the class, with some thoughts.

https://github.com/WordPress/gutenberg/pull/46579

Reply:
Good catch. Done!